src/backend/cdb/dispatcher/cdbdisp_query.c

```diff
@@ -75,6 +76,7 @@ typedef struct DispatchCommandQueryParms
 	 */
 	const char *strCommand;
 	int			strCommandlen;
+	int			queryCommandId;
```

Comment:
Why do we need this new field? The tests pass without it.
```diff
diff --git a/src/backend/cdb/dispatcher/cdbdisp_query.c b/src/backend/cdb/dispatcher/cdbdisp_query.c
index 07266234a4e..bb901a603aa 100644
--- a/src/backend/cdb/dispatcher/cdbdisp_query.c
+++ b/src/backend/cdb/dispatcher/cdbdisp_query.c
@@ -76,7 +76,6 @@ typedef struct DispatchCommandQueryParms
 	 */
 	const char *strCommand;
 	int			strCommandlen;
-	int			queryCommandId;
 	char	   *serializedQuerytree;
 	int			serializedQuerytreelen;
 	char	   *serializedPlantree;
@@ -528,7 +527,6 @@ cdbdisp_buildCommandQueryParms(const char *strCommand, int flags)
 
 	pQueryParms = palloc0(sizeof(*pQueryParms));
 	pQueryParms->strCommand = strCommand;
-	pQueryParms->queryCommandId = MyProc->queryCommandId;
 	pQueryParms->serializedQuerytree = NULL;
 	pQueryParms->serializedQuerytreelen = 0;
 	pQueryParms->serializedQueryDispatchDesc = NULL;
@@ -602,7 +600,6 @@ cdbdisp_buildUtilityQueryParms(struct Node *stmt,
 
 	pQueryParms = palloc0(sizeof(*pQueryParms));
 	pQueryParms->strCommand = PointerIsValid(debug_query_string) ? debug_query_string : "";
-	pQueryParms->queryCommandId = MyProc->queryCommandId;
 	pQueryParms->serializedQuerytree = serializedQuerytree;
 	pQueryParms->serializedQuerytreelen = serializedQuerytree_len;
 	pQueryParms->serializedQueryDispatchDesc = serializedQueryDispatchDesc;
@@ -685,7 +682,6 @@ cdbdisp_buildPlanQueryParms(struct QueryDesc *queryDesc,
 	sddesc = serializeNode((Node *) queryDesc->ddesc, &sddesc_len, NULL /* uncompressed_size */ );
 
 	pQueryParms->strCommand = queryDesc->sourceText;
-	pQueryParms->queryCommandId = MyProc->queryCommandId;
 	pQueryParms->serializedQuerytree = NULL;
 	pQueryParms->serializedQuerytreelen = 0;
 	pQueryParms->serializedPlantree = splan;
@@ -895,7 +891,7 @@ buildGpQueryString(DispatchCommandQueryParms *pQueryParms,
 	const char *dtxContextInfo = pQueryParms->serializedDtxContextInfo;
 	int			dtxContextInfo_len = pQueryParms->serializedDtxContextInfolen;
 	int64		currentStatementStartTimestamp = GetCurrentStatementStartTimestamp();
-	int			queryCommandId = pQueryParms->queryCommandId;
+	int			queryCommandId = MyProc->queryCommandId;
 	Oid			sessionUserId = GetSessionUserId();
 	Oid			outerUserId = GetOuterUserId();
 	Oid			currentUserId = GetUserId();
```

Reply:
Yes, it can work without the field. But it looks more consistent to pass the value in DispatchCommandQueryParms structure together with other properties of the query. That is why we assign the value in `cdbdisp_build...` function instead of directly assigning it in `buildGpQueryString()`

Reply:
`numsegments` is also directly set in the `buildGpQueryString` function.

Reply:
`numsegments` is not coupled to the query, as the query id is.